subscription.go

```diff
@@ -31,3 +47,33 @@ func (sub *Subscription) Next(ctx context.Context) (*Message, error) {
 func (sub *Subscription) Cancel() {
 	sub.cancelCh <- sub
 }
+
+func (sub *Subscription) close() {
+	close(sub.ch)
+	close(sub.joinCh)
+	close(sub.leaveCh)
+}
+
+// NextPeerEvent returns the next event regarding subscribed peers
+// Note: There is no guarantee that the Peer Join event will fire before
+// the related Peer Leave event for a given peer
```

Comment:
Should note that this is lossy.

Reply:
Actually, this doesn't even need to be lossy. We can have this function:

1. Return all connected peers on start.
2. Guarantee that it returns all connect/disconnect events.

We can do this by keeping track of peers we've reported as subscribed internally. If we fall behind, pubsub can close the event channel. If we see that the event channel has been closed, we can re-register for events and, at the same time, diff our local list of tracked peers against the master list.

---

Also, we really should guarantee ordering of these events. Is that not possible?

Reply:
Good idea. The reason we couldn't guarantee ordering of events was because we didn't want the `PeerLeave` events filling up the event queue to block `PeerJoin` events from making it in so we had to process them independently before we could combine them at the output.

However, your scheme should allow us to do both and the only tradeoff is that we won't see peers that quickly Join and Leave, but that feels pretty irrelevant.

Reply:
sounds good to me!

Reply:
@vyzo @Stebalien did this, need an updated review. I'd like more eyes to confirm my strategy isn't racy, although I added some tests to try and help a bit with that.